tools/python/util/onnx_model_utils.py

```diff
@@ -0,0 +1,311 @@
+                if dim.HasField('dim_value'):
+                    if dim.dim_value != fixed_shape[idx]:
+                        raise ValueError(
+                            f"Can't replace existing fixed size of {dim.dim_value} with {fixed_shape[idx]} "
+                            f"for dimension {idx + 1}")
+                elif dim.HasField('dim_param'):
+                    # replacing a dim_param so have to do that through the entire graph
+                    make_dim_param_fixed(graph, dim.dim_param, fixed_shape[idx])
+                else:
+                    # replacing an unknown dim
+                    dim.Clear()
+                    dim.dim_value = fixed_shape[idx]
+
+                idx += 1
+
+            return
+
+    raise ValueError(f'Input {input_name} was not found in graph inputs. '
+                     f'Valid input names are: {",".join([i.name for i in graph.input])}')
+
+
+def _create_producer_consumer_link(node_to_producers: dict, node_to_consumers: dict,
+                                   producer: onnx.NodeProto, consumer: onnx.NodeProto):
+    '''
+    Create links between two nodes for a value produced by one and consumed by the other.
+    :param node_to_producers: Map of NodeProto to set of nodes that produce values the node consumes as inputs.
+    :param node_to_consumers: Map of NodeProto to set of nodes that consume values the node produces as outputs.
+    :param producer: Producer node
+    :param consumer: Consumer node
+    '''
+
+    if consumer not in node_to_producers:
+        node_to_producers[consumer] = set()
+
+    if producer not in node_to_consumers:
+        node_to_consumers[producer] = set()
+
+    # add entry mapping this node to the producer of this input
+    node_to_producers[consumer].add(producer)
+    node_to_consumers[producer].add(consumer)
+
+
+def _map_node_dependencies(graph: onnx.GraphProto, node_to_producers: dict, node_to_consumers: dict):
+    graph_inputs = set([i.name for i in graph.input])
+    initializers = set([i.name for i in graph.initializer])
+
+    # map of value name to node that creates it. copy parent values but override if values get shadowed
+    producers = {}
+
+    implicit_inputs = set()
+
+    def is_local_value(value):
+        return value in producers or value in initializers or value in graph_inputs
+
+    for node in graph.node:
+        inputs = [i for i in node.input]
+
+        for attr in node.attribute:
+            if attr.HasField('g'):
+                subgraph_implicit_inputs = _map_node_dependencies(attr.g, node_to_producers, node_to_consumers)
+                inputs += subgraph_implicit_inputs
+
+        for i in inputs:
+            if not i:
+                # missing optional input
+                continue
+
+            if is_local_value(i):
+                if i in producers:
+                    producer = producers[i]
+                    _create_producer_consumer_link(node_to_producers, node_to_consumers, producer, node)
```

Comment:
>_create_producer_consumer_link [](http://example.com/codeflow?start=20&length=30)

is a link created between a producer and a consumer in a subgraph?

Reply:
I see, linking to the node with the subgraph makes sense too. This seems fine to me, I was just curious about how that case was handled.